### PR TITLE
FEATURE: mailpace webhook

### PR DESCRIPTION
**Pull request by ruaq**
adds mailpace (formerly known as ohmystmp) webhook

### CODE REVIEW

File: app/controllers/webhooks_controller.rb [hunk 1]
```diff
@@ -48,6 +48,23 @@ def mailjet
     success
   end
 
+  def mailpace
+    # see https://docs.mailpace.com/guide/webhooks#email-events
+
+    message_id = Email::MessageIdService.message_id_clean(params["message_id"])
+    to_address = params["to"]
+    status = params["Type"]
```
Comment (nbianca) [inline app/controllers/webhooks_controller.rb:56]: Are you sure this is the right key? According to the documentation you linked in the comment above, I don't see any `Type` key, but a `status` one. Is that what you meant?

File: spec/requests/webhooks_controller_spec.rb [hunk 2]
```diff
@@ -126,6 +126,49 @@
...
+        }
+      }
+
+      expect(response.status).to eq(200)
+
+      email_log.reload
+      expect(email_log.bounced).to eq(true)
+      expect(email_log.bounce_error_code).to eq(nil) # mailpace doesn't give us this
+      expect(email_log.user.user_stat.bounce_score).to eq(SiteSetting.hard_bounce_score)
+    end
```
Comment (nbianca) [inline spec/requests/webhooks_controller_spec.rb:149]:
```suggestion
    end

```

File: app/controllers/webhooks_controller.rb (hunk 1 above)
Comment (ruaq) [reply app/controllers/webhooks_controller.rb:56]: You're right. Crap. I'm used to using PHPStorm which autosave. I used VSCodium this time. I've changed it locally but haven't saved this change. thx for the hint!

**Comment by nbianca** [comment]
The tests fail. Can you please have a look?

Thank you for your contribution!

**Comment by ruaq** [comment]
My Mac _died_ a few days ago. So it's a little bit tricky at the moment. But I think with the last change it should pass the tests. Would be great if you can give it another run, @nbianca

File: spec/requests/webhooks_controller_spec.rb [hunk 3]
```diff
@@ -126,6 +126,50 @@
...
+      email_log = Fabricate(:email_log, user: user, message_id: message_id, to_address: email)
+
+      post "/webhooks/mailpace.json", params: {
+        "event": "email.bounced",
+        "payload": [{
+            "status": "bounced",
+            "to": email,
+            "message_id": "<#{message_id}>",
+        }]
+      }
```
Comment (paul-oms) [inline spec/requests/webhooks_controller_spec.rb:141]: This should be:

```
      post "/webhooks/mailpace.json", params: {
        "event": "email.bounced",
        "payload": {
            "status": "bounced",
            "to": email,
            "message_id": "<#{message_id}>",
        }
      }
```

File: spec/requests/webhooks_controller_spec.rb [hunk 4]
```diff
@@ -126,6 +126,50 @@
...
+      email_log = Fabricate(:email_log, user: user, message_id: message_id, to_address: email)
+
+      post "/webhooks/mailpace.json", params: {
+        "event": "email.deferred",
+        "payload": [{
+            "status": "deferred",
+            "to": email,
+            "message_id": "<#{message_id}>",
+        }]
+      }
```
Comment (paul-oms) [inline spec/requests/webhooks_controller_spec.rb:162]: This should be:
```

      post "/webhooks/mailpace.json", params: {
        "event": "email.bounced",
        "payload": {
            "status": "bounced",
            "to": email,
            "message_id": "<#{message_id}>",
        }
      }
```

File: app/controllers/webhooks_controller.rb [hunk 5]
```diff
@@ -48,6 +48,23 @@ def mailjet
     success
   end
 
+  def mailpace
+    # see https://docs.mailpace.com/guide/webhooks#email-events
+
+    message_id = Email::MessageIdService.message_id_clean(params["message_id"])
+    to_address = params["to"]
+    status = params["status"]
```
Comment (paul-oms) [inline app/controllers/webhooks_controller.rb:56]: This should be:

```
    message_id = Email::MessageIdService.message_id_clean(params[:payload][:message_id])
    to_address = params[:payload][:to]
    status = params[:payload][:status]
```

**Comment by paul-oms** [comment]
Hi @ruaq @nbianca  - creator of mailpace.com here. I've added some comments on this PR that will fix this if implemented. If I can be granted the right permissions I can push up the changes in a single commit (tests pass locally for me)

**Comment by ruaq** [comment]
I’ve changed it to your suggestions, @paul-oms

Thanks so much for your support.

**Comment by pmusaraj** [comment]
@nbianca do you have the time to follow up on this PR?

File: app/controllers/webhooks_controller.rb [hunk 6]
```diff
@@ -48,6 +48,23 @@ def mailjet
     success
   end
 
+  def mailpace
```
Comment (nbianca) [inline app/controllers/webhooks_controller.rb:51]: This is missing any webhook verification, but I believe we can merge it now and improve this later.

**Comment by nattsw** [comment]
Lint problem in the spec and spec failure:

```
  1) WebhooksController#mailpace soft bounces
     Failure/Error: expect(email_log.user.user_stat.bounce_score).to eq(SiteSetting.soft_bounce_score)

       expected: 1
            got: 2.0

       (compared using ==)
     # ./spec/requests/webhooks_controller_spec.rb:254:in `block (3 levels) in <main>'
```

@ruaq think you could have a look?

File: app/controllers/webhooks_controller.rb [hunk 7]
```diff
@@ -48,6 +48,23 @@ def mailjet
...
+    # see https://docs.mailpace.com/guide/webhooks#email-events
+
+    message_id = Email::MessageIdService.message_id_clean(params[:payload][:message_id])
+    to_address = params[:payload][:to]
+    status = params[:payload][:status]
```
Comment (nbianca) [inline app/controllers/webhooks_controller.rb:56]:
```suggestion
    message_id = Email::MessageIdService.message_id_clean(params["payload"]["message_id"])
    to_address = params["payload"]["to"]
    status = params["payload"]["status"]
```

File: spec/requests/webhooks_controller_spec.rb [hunk 8]
```diff
@@ -129,6 +129,51 @@
...
+      email_log = Fabricate(:email_log, user: user, message_id: message_id, to_address: email)
+
+      post "/webhooks/mailpace.json", params: {
+        "event": "email.bounced",
+        "payload": {
+            "status": "bounced",
+            "to": email,
+            "message_id": "<#{message_id}>",
+        }
+      }
```
Comment (nbianca) [inline spec/requests/webhooks_controller_spec.rb:144]:
```suggestion
      post "/webhooks/mailpace.json",
           params: {
             event: "email.bounced",
             payload: {
               status: "bounced",
               to: email,
               message_id: "<#{message_id}>",
             },
           }
```

**Pull request closed without merging**